src/native/fs/fs_napi.cpp

```diff
@@ -0,0 +1,747 @@
+/* Copyright (C) 2016 NooBaa */
+#include "../util/b64.h"
+#include "../util/common.h"
+#include "../util/napi.h"
+
+#include <uv.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <fcntl.h>
+#include <unistd.h>
+#include <dirent.h>
+#include <vector>
+// #include <errno.h>
+// #include <sys/fsuid.h>
+
+namespace noobaa
+{
+
+struct Entry {
+    std::string name;
+    ino_t ino;
+    uint8_t type;
+};
+
+template <typename T>
+static Napi::Value api(const Napi::CallbackInfo& info)
+{
+    auto w = new T(info);
+    Napi::Promise promise = w->_deferred.Promise();
+    w->Queue();
+    return promise;
+}
+
+/**
+ * FSWorker is a general async worker for our fs operations
+ */
+struct FSWorker : public Napi::AsyncWorker
+{
+    Napi::Promise::Deferred _deferred;
+    int _errno;
+    std::string _desc;
+
+    FSWorker(const Napi::CallbackInfo& info)
+        : AsyncWorker(info.Env())
+        , _deferred(Napi::Promise::Deferred::New(info.Env()))
+        , _errno(0)
+    {
+    }
+    void Begin(std::string desc)
+    {
+        _desc = desc;
+        LOG("FS::FSWorker::Begin: " << _desc);
+    }
+    void SetSyscallError()
+    {
+        _errno = errno;
```

Comment:
I think we want to only set it if _errno is not set (0) otherwise log and return.